frontend/pre_analysis.py

```diff
@@ -14,9 +16,162 @@ def __init__(self, prog_ast):
     def maximum_function_args(self):
         """Get the maximum number of function arguments appearing in the AST"""
         func_defs = [node for node in self.all_nodes if isinstance(node, ast.FunctionDef)]
-        return 0 if len(func_defs) == 0 else max([len(node.args.args) for node in func_defs])
+        return 1 if not func_defs else max([len(node.args.args) for node in func_defs])
 
     def maximum_tuple_length(self):
         """Get the maximum length of tuples appearing in the AST"""
         tuples = [node for node in self.all_nodes if isinstance(node, ast.Tuple)]
-        return 0 if len(tuples) == 0 else max([len(node.elts) for node in tuples])
+        return 0 if not tuples else max([len(node.elts) for node in tuples])
+
+    def analyze_classes(self):
+        """Pre-analyze and configure classes before the type inference
+        
+        Do the following:
+            - Propagate methods from bases to subclasses.
+            - Add __init__ function to classes if it doesn't exist.
+            - Return a mapping from class names to their attributes and methods.
+            - Return a mapping from class names to their base classes if they have any.
+            
+        """
+        # TODO propagate attributes to subclasses.
```

Comment:
Methods are defined by normal function definitions, and methods are propagated by copying their `FunctionDef` AST nodes to the subclasses. However, attributes work differently. They can be defined through normal assignments as class-level variables, or through attribute assignment on the first argument in the methods.

For example, below is two valid ways to define two attributes `x` and `y`:

```python
class A:
    x = 1
    def f(self):
        self.y = ""
```

So a different mechanism will be used to propagate attributes.